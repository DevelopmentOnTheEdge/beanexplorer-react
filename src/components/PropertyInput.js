import React                from 'react';
import PropTypes            from 'prop-types';
import Datetime             from 'react-datetime';
import moment               from 'moment';
import Select               from 'react-select';
import VirtualizedSelect    from 'react-virtualized-select'
import NumericInput          from 'react-numeric-input';
import CKEditor              from 'react-ckeditor-component';
import MaskedInput           from 'react-maskedinput';
import JsonPointer          from 'json-pointer';


class PropertyInput extends React.Component
{
  constructor(props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
  }

  getPath() {
    if(this.props.path) {
      return this.props.path;
    }else{
      return this.props.bean.order[this.props.id];
    }
  }

  callOnChange(value) {
    this.props.onChange(this.getPath(), value);
  }

  handleChange(event) {
    const value = (event.target.type === 'checkbox') ? event.target.checked : event.target.value;
    this.callOnChange(value);
  }

  dateToISOFormat(date) {
    if(typeof date === "string") {
      this.callOnChange(date);
    } else {
      this.callOnChange(date.format('YYYY-MM-DD'));
    }
  }

  dateFromISOFormat(stringDate) {
    const date = moment(stringDate === undefined ? "" : stringDate, 'YYYY-MM-DD', true);
    if (date.isValid()) {
      return date.format('DD.MM.YYYY');
    } else {
      return stringDate;
    }
  }

  handleChangeSelect(object) {
    if(Array.isArray(object)) {
      let selectArray = [];
      Object.keys(object).forEach(function (key) {
        selectArray.push(object[key].value);
      });
      this.callOnChange(selectArray);
    } else {
      this.callOnChange(object !== null ? object.value : "");
    }
  }

  static getMaskInput(rules)
  {
    for (let i = 0; i < rules.length; i++)
    {
      if ("mask" in rules[i]) {
        return rules[i].mask
      }
    }
    return null;
  }

  static isNumberInput(rules)
  {
    for (let i =0 ; i< rules.length; i++)
    {
      if(rules[i].type === "baseRule" &&
        ( rules[i].attr === "digits" || rules[i].attr === "integer" || rules[i].attr === "number" ))return true;
    }
    return false;
  }

  static getNumericProps(meta)
  {
    let props = {};
    props['maxLength'] = 14;//errors if more
    const rules = meta.validationRules;
    for (let i =0 ; i< rules.length; i++)
    {
      if(rules[i].type === "baseRule" && (rules[i].attr === "number"))
      {
        props['precision'] = 10;
      }
      if(rules[i].type === "baseRule" && (rules[i].attr === "integer"))
      {
        props['min'] = -2147483648;
        props['max'] = 2147483647;
        props['maxLength'] = 9;
        props['precision'] = 0;
      }
      // if(rules[i].type === "digits")
      // {
      //   props['min'] = 0;//todo not work
      // }
    }
    if(meta.columnSize){
      props['maxLength'] = parseInt(meta.columnSize);
    }
    return props;
  }

  static getBase64(file) {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.addEventListener("load", function () {
        resolve(reader.result);
      }, false);
      reader.onerror = error => reject(error);

      reader.readAsDataURL(file);
    });
  }

  static getExtraAttrsMap(extraAttrs){
    let map = {};
    if(extraAttrs === undefined)return map;
    for (let i=0 ;i< extraAttrs.length; i++){
      map[extraAttrs[i][0]] = extraAttrs[i][1];
    }
    return map;
  }

  render() {
    const path  = this.getPath();
    const meta  = this.props.bean.meta[path];
    const value = JsonPointer.get(this.props.bean, "/values" + path);
    const id    = path.substring(path.lastIndexOf("/")+1) + "PropertyInput";
    const extraAttrsMap = PropertyInput.getExtraAttrsMap(meta.extraAttrs);
    const required = meta.canBeNull !== true;

    const baseProps = {
      id: id,
      key: id,
      disabled: meta.readOnly,
      required: required
    };

    const rawTextInputProps = Object.assign({}, baseProps, {
      value: value === undefined ? "" : value,
      onChange: this.handleChange,
      placeholder: meta.placeholder,
      className: this.props.controlClassName || "form-control"
    });

    const controls = {
      textInput: () => (
        <input type="text" {...rawTextInputProps} />
      ),
      passwordField: () => (
        <input type="password" {...rawTextInputProps} />
      ),
      textArea: () => (
        <textarea rows={meta.rows || 3} cols={meta.columns} {...rawTextInputProps} />
      ),
      Boolean: () => (
        <input type="checkbox" checked={value === true || value === "true"} onChange={this.handleChange}
               className={this.props.controlClassName || 'form-check-input'} {...baseProps} />
      ),
      file: () => (
        <input type="file" className={this.props.controlClassName || "form-control-file"} {...baseProps}
                      multiple={meta.multipleSelectionList}
                      onChange={(e) => {
                        if(e.target.files && e.target.files.length === 1) {
                          const fileName = e.target.files[0].name;
                          PropertyInput.getBase64(e.target.files[0]).then(data => {
                            this.callOnChange({type: "Base64File", name: fileName, data: data})
                          });
                        }else if(e.target.files && e.target.files.length === 0) {
                          this.callOnChange("")
                        }
                      }} />
      ),
      select: () => {
        let options = [];
        for(let i =0 ;i < meta.tagList.length; i++){
          options.push({ value: meta.tagList[i][0], label: meta.tagList[i][1] });
        }

        let strValue;
        if(Array.isArray(value)){
          strValue = [];
          for (let i = 0; i < value.length; i++)strValue.push("" + value[i]);
        }
        else
        {
          strValue = "" + value;
        }
        const selectAttr = {
          ref: id, name: id, value: strValue, options: options, onChange: (v) => this.handleChangeSelect(v),
          clearAllText: this.props.localization.clearAllText,
          clearValueText: this.props.localization.clearValueText,
          noResultsText: this.props.localization.noResultsText,
          searchPromptText: this.props.localization.searchPromptText,
          loadingPlaceholder: this.props.localization.loadingPlaceholder,
          placeholder: meta.placeholder || this.props.localization.placeholder,
          backspaceRemoves: false,
          disabled: meta.readOnly,
          multi: meta.multipleSelectionList,
          matchPos: extraAttrsMap.matchPos || "any",
          required: required
        };

        if(extraAttrsMap.inputType === "Creatable")
        {
          return <Creatable {...selectAttr} />
        }
        else if(extraAttrsMap.inputType === "VirtualizedSelect")
        {
          return <VirtualizedSelect {...selectAttr} clearable searchable labelKey="label" valueKey="value" />
        }
        else
        {
          return <Select {...selectAttr} />
        }
      },
      Date: () => (
        <Datetime dateFormat="DD.MM.YYYY" id={id} key={id} inputProps={ {disabled: meta.readOnly, required: required} }
                         onChange={(v) => this.dateToISOFormat(v)} value={this.dateFromISOFormat(value)}
                         timeFormat={false} closeOnSelect={true} closeOnTab={true} locale={this.props.localization.locale || "en"} />
      ),
//      dateTime: {
//        normal: () => {
//          return ( React.createElement(Datetime, {id: id, key: id, value: value, parent: _this, onChange: handleChange, time: true, className: this.props.controlClassName}) );
//        },
//        readOnly: () => this.createStatic(value)
//      },
      maskTest: () => (
        <MaskedInput mask={PropertyInput.getMaskInput(meta.validationRules)} value={value === undefined ? "" : value}
                            onChange={this.handleChange} className={this.props.controlClassName || "form-control"} {...baseProps} />
      ),
      numberInput: () => {
        const numericProps = PropertyInput.getNumericProps(meta);
        return <NumericInput {...numericProps} placeholder={meta.placeholder} value={value} {...baseProps}
                             onChange={(valueAsNumber, valueAsString, input) => {
                               this.callOnChange(valueAsNumber !== null ? valueAsNumber : "");
                             }}
                             style={ false } className={this.props.controlClassName || "form-control"} />
      },
      WYSIWYG: () => (
        <CKEditor activeClass="p10" content={value}
                         events={{
                           "change": (evt) => { this.callOnChange(evt.editor.getData()) }
                         }}
                         config={{language: 'ru', readOnly: meta.readOnly}}
        />
      ),
      labelField: () => {
        if(meta.rawValue)
        {
          return <div dangerouslySetInnerHTML={{__html: value}} />
        }
        else
        {
          return <label className="form-control-label">{value}</label>
        }
      },
    };

    if(meta.tagList)
    {
      return controls['select']();
    }

    if(meta.passwordField)
    {
      return controls['passwordField']();
    }

    if(meta.labelField)
    {
      return controls['labelField']();
    }

    if(meta.validationRules !== undefined && PropertyInput.isNumberInput(meta.validationRules))
    {
      return controls['numberInput']()
    }

    if(extraAttrsMap.inputType === 'WYSIWYG')
    {
      return controls['WYSIWYG']();
    }

    if(extraAttrsMap.inputType === 'textArea')
    {
      return controls['textArea']();
    }

    if(extraAttrsMap.inputType === 'file')
    {
      return controls['file']();
    }

    if(meta.validationRules !== undefined && PropertyInput.getMaskInput(meta.validationRules))
    {
      return controls['maskTest']();
    }

    if(controls[meta.type] !== undefined)
    {
      return controls[meta.type]();
    }

    return controls['textInput']();
  }

}

PropertyInput.defaultProps = {
  localization: {
    locale: 'en',
    clearAllText: 'Clear all',
    clearValueText: 'Clear value',
    noResultsText: 'No results found',
    searchPromptText: 'Type to search',
    placeholder: 'Select ...',
    loadingPlaceholder: 'Loading...'
  },
};

PropertyInput.propTypes = {
  bean: PropTypes.object.isRequired,
  path: PropTypes.string,
  id: PropTypes.number,
  onChange: PropTypes.func,
  localization: PropTypes.object,
  controlClassName: PropTypes.string, 
};

export default PropertyInput;
