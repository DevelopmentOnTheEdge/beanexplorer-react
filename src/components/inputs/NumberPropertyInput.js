import React from 'react';
import bigInt from "big-integer";
import bigRat from "big-rational";
import BasePropertyInput from "./BasePropertyInput";


export default class NumberPropertyInput extends BasePropertyInput {
  constructor(props) {
    super(props);
    this.numberValidation = this.numberValidation.bind(this);
  }

  render() {
    const meta = this.getMeta();
    const extraAttrsMap = BasePropertyInput.getExtraAttrsMap(meta);
    const range = this.getNumberValidationRule('range');
    const step = this.getNumberValidationRule('step');
    const type = meta.type;

    return <input
      type="text"
      onInput={this.numberValidation}
      data-info-type={type}
      data-info-range={(range && range.attr) ? range.attr.min + ', ' + range.attr.max : undefined}
      data-info-step={step ? step.attr : undefined}
      value={convertENotationNumbers(this.getValue())}
      onChange={this.handleChange}
      placeholder={extraAttrsMap.placeholder}
      {...this.getBaseProps()}
    />
  }

  numberValidation(e) {
    const range = this.getNumberValidationRule('range');
    const step = this.getNumberValidationRule('step');

    const local = this.props.localization;

    let value;
    try {
      value = bigRat(e.target.value);
    } catch (err) {
      setErrorState(e, local.numberTypeMismatch);
      return
    }

    if (range) {
      if (value.compare(bigRat(range.attr.min)) === -1) {
        setErrorState(e, setMessagePlaceHolders(local.rangeUnderflow, [range.attr.min]));
        return
      }
      else if (value.compare(bigRat(range.attr.max)) === 1) {
        setErrorState(e, setMessagePlaceHolders(local.rangeOverflow, [range.attr.max]));
        return
      }
    }

    if (step) {
      const stepRat = bigRat(step.attr);

      if (!value.divide(stepRat).denominator.equals(bigInt.one)) {
        const min = value.divide(stepRat).floor().multiply(stepRat);
        const max = min.add(stepRat);

        setErrorState(e, setMessagePlaceHolders(
          local.stepMismatch, [min.toDecimal(), max.toDecimal()]
        ));
        return
      }
    }

    setErrorState(e, '');
  }

  getNumberValidationRule(name) {
    const meta = this.getMeta();
    let rule = this.getValidationRule(name);

    if (rule === undefined)
    {
      if (name === 'range') {
        switch (meta.type) {
          case 'Short':
            return {type: 'range', attr: {min: "-32768", max: "32767"}};
          case 'Integer':
            return {type: 'range', attr: {min: "-2147483648", max: "2147483647"}};
          case 'Long':
            return {type: 'range', attr: {min: "-9223372036854775808", max: "9223372036854775807"}};
        }
      }
      if (name === 'step' && (meta.type === 'Short' || meta.type === 'Integer' || meta.type === 'Long')) {
        return {type: 'step', attr: '1'};
      }
    }
    return rule;
  };
}

const convertENotationNumbers = function (value) {
  try {
    if (value.includes('e') || value.includes('E'))
      return bigRat(value).toDecimal();
    else
      return value;
  } catch (err) {
    return value
  }
};

const setErrorState = function (e, text) {
  e.target.setCustomValidity(text);
  e.target.title = text;
};

const setMessagePlaceHolders = function (source, params) {
  params.forEach(function (item, i) {
    source = source.replace(new RegExp("\\{" + i + "\\}", "g"), item);
  });
  return source
};
