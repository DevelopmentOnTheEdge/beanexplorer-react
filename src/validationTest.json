{
  "values": {
    "short": 32767,
    "integer": 2147483647,
    "long": 9007199254740991,
    "double": 100.1,
    "doubleWithStep": 100.5,
    "integerWithRangeAndStep": "",
    "textInput": "qwerty7890"
  },
  "meta": {
    "/short": {
      "displayName": "Short",
      "type": "Short",
      "groupName": "Numbers",
      "groupId": 1
    },
    "/integer": {
      "displayName": "Integer",
      "type": "Integer",
      "groupId": 1
    },
    "/long": {
      "displayName": "Long",
      "type": "Long",
      "message": "use Number.MAX_SAFE_INTEGER and Number.MIN_SAFE_INTEGER",
      "groupId": 1
    },
    "/double": {
      "displayName": "Double",
      "type": "Double",
      "message": "have problems in firefox",
      "groupId": 1
    },
    "/doubleWithStep": {
      "displayName": "Double with custom step",
      "type": "Double",
      "validationRules":[{"type":"step","attr":0.5},{"type":"range","attr":{"min":0,"max":1000}}],
      "message": "Use also the 'range' for correct work in chrome, firefox, etc",
      "groupId": 1
    },
    "/integerWithRangeAndStep": {
      "displayName": "Number input with range and integer step",
      "validationRules":[{"type":"step","attr":10},{"type":"range","attr":{"min":0,"max":100}}],
      "groupId": 1,
      "canBeNull": true
    },
    "/textInput": {
      "displayName": "Text input with columnSize",
      "columnSize": 10,
      "groupName": "Text",
      "groupId": 2
    }
  },
  "order": [
    "/short",
    "/integer",
    "/long",
    "/double",
    "/doubleWithStep",
    "/integerWithRangeAndStep",
    "/textInput"
  ]
}