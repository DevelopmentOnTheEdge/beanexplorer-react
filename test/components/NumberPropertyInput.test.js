import React from 'react';
import PropertyInput from '../../src/components/PropertyInput';
import {shallow, mount, render} from 'enzyme';
import bean from '../../src/testJson.json';
import validationTest from '../../src/validationTest.json';
import NumberPropertyInput from "../../src/components/inputs/NumberPropertyInput";
import renderer from 'react-test-renderer';


test('textInputWithPatternAndMessage numbers', () => {
  const handle = jest.fn();

  const spy = jest.spyOn(NumberPropertyInput.prototype, "numberValidation");
  const wrapper = mount(
    <PropertyInput path={"/number"} bean={bean} value={"123"}/>
  );

  wrapper.find('input').simulate('input', {target: {value: "1", setCustomValidity: handle}});
  expect(spy.mock.calls.length).toEqual(1);
  expect(handle.mock.calls[0]).toEqual([""]);

  wrapper.find('input').simulate('input', {target: {value: "", setCustomValidity: handle}});
  expect(handle.mock.calls[1]).toEqual([""]);

  wrapper.find('input').simulate('input', {target: {value: "a", setCustomValidity: handle}});
  expect(handle.mock.calls[2]).toEqual(["Enter the number."]);

  wrapper.find('input').simulate('input', {target: {value: "2147483648", setCustomValidity: handle}});
  expect(handle.mock.calls[3]).toEqual(["The value must be less than or equal to 2147483647."]);

  wrapper.find('input').simulate('input', {target: {value: "-2147483649", setCustomValidity: handle}});
  expect(handle.mock.calls[4]).toEqual(["The value must be greater than or equal to -2147483648."]);

  wrapper.find('input').simulate('input', {target: {value: "1.1", setCustomValidity: handle}});
  expect(handle.mock.calls[5]).toEqual(["Please enter a valid value. The closest allowed values are 1 and 2."]);

  wrapper.setProps({bean: validationTest, path: "/short"});

  ["/short","/integer","/long","/double","/doubleWithRangeAndStep","/integerWithRangeAndStep",].forEach(function(path) {
    wrapper.setProps({path: path});
    wrapper.find('input').simulate('input', {target: {value: "1", setCustomValidity: handle}});
  });
  expect(spy.mock.calls.length).toEqual(12);
  expect(handle.mock.calls.length).toEqual(12);
});

test('convertENotationNumbers', () => {
  const bean = {
    "values": { "double": "1.1e2" },
    "meta":   { "/double": {"type": "Double"} },
    "order":  [ "/double" ]
  };

  const component = renderer.create(
    <PropertyInput path={"/double"} bean={bean} value={"1.1e2"}/>
  );
  expect(component.toJSON()).toMatchSnapshot();

  const component2 = renderer.create(
    <PropertyInput path={"/double"} bean={bean} value={"text"}/>
  );
  expect(component2.toJSON()).toMatchSnapshot();
});
