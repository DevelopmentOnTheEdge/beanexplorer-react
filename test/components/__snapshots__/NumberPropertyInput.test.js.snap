// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertENotationNumbers 1`] = `
<input
  className="property-input form-control"
  data-info-type="Double"
  id="doublePropertyInput"
  onBlur={[Function]}
  onChange={[Function]}
  onInput={[Function]}
  required={true}
  type="text"
  value="110"
/>
`;

exports[`convertENotationNumbers 2`] = `
<input
  className="property-input form-control"
  data-info-type="Double"
  id="doublePropertyInput"
  onBlur={[Function]}
  onChange={[Function]}
  onInput={[Function]}
  required={true}
  type="text"
  value="text"
/>
`;
